pkg/controller/signer/signer-controller.go

```diff
@@ -0,0 +1,269 @@
+package signer
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/openshift/cluster-network-operator/pkg/controller/statusmanager"
+	"github.com/openshift/library-go/pkg/crypto"
+	csrv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/client-go/kubernetes"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const signerName = "network.openshift.io/signer"
+
+// Add controller and start it when the Manager is started.
+func Add(mgr manager.Manager, status *statusmanager.StatusManager) error {
+	reconciler, err := newReconciler(mgr, status)
+	if err != nil {
+		return err
+	}
+	return add(mgr, reconciler)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager) (reconcile.Reconciler, error) {
+	// We need a clientset in order to UpdateApproval() of the CertificateSigningRequest
+	clientset, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, err
+	}
+	return &ReconcileCSR{client: mgr.GetClient(), scheme: mgr.GetScheme(), status: status, clientset: clientset}, nil
+}
+
+// This controller will not handle certificate updates
+func ignoreUpdatePredicate() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("signer-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to CetificateSigningRequest resource
+	err = c.Watch(&source.Kind{Type: &csrv1.CertificateSigningRequest{}}, &handler.EnqueueRequestForObject{},
+		ignoreUpdatePredicate())
```

Comment:
I wouldn't do this - controllers **must** be level-triggered.

Reply:
Then why does the controller-runtime allow for it?

Reply:
No clue, but it's definitely not recommended.

Besides, you'll get an `ADD` event every time the process is restarted, regardless of the state of the object.